huggingface_pipelines/pipeline.py

```diff
@@ -0,0 +1,221 @@
+
+        Returns:
+            PipelineConfig: A new PipelineConfig instance with the applied overwrites.
+        """
+        return replace(self, **overwrites)
+
+
+class Pipeline(ABC):
+    """
+    Abstract base class for different pipelines.
+
+    This class defines the common structure and methods for all pipeline implementations.
+    Specific pipeline classes should inherit from this class and implement the
+    `process_batch` method.
+
+    Attributes:
+        config (PipelineConfig): The configuration for this pipeline.
+    """
+
+    def __init__(self, config: PipelineConfig):
+        """
+        Initialize the Pipeline with the given configuration.
+
+        Args:
+            config (PipelineConfig): The configuration for this pipeline.
+        """
+        self.config = config
+
+    @contextmanager
+    def resource_manager(self):
+        """
+        Context manager to efficiently initialize and free pipeline resources.
+
+        This method ensures that CUDA memory is properly managed when using GPU.
+        It clears the CUDA cache before and after pipeline execution.
+
+        Yields:
+            None
+        """
+        try:
+            if torch.cuda.is_available() and self.config.device == 'cuda':
+                torch.cuda.empty_cache()
+            yield
+        finally:
+            if torch.cuda.is_available() and self.config.device == 'cuda':
+                torch.cuda.empty_cache()
+
+    @abstractmethod
+    def process_batch(self, batch: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        Processes a single batch of data and returns the updated batch.
+
+        This method should be implemented by all concrete pipeline classes.
+
+        Args:
+            batch (Dict[str, Any]): A batch of data to process.
+
+        Returns:
+            Dict[str, Any]: The processed batch of data.
+        """
+        pass
+
+    def __call__(self, dataset: Dataset) -> Dataset:
+        """
+        Processes the entire dataset using the pipeline.
+
+        This method applies the `process_batch` method to the entire dataset,
+        handling batching, caching, and error management.
+
+        Args:
+            dataset (Dataset): The dataset to process.
+
+        Returns:
+            Dataset: The processed dataset.
+
+        Raises:
+            Exception: If there's an error during dataset processing.
+        """
+        try:
+            logger.info("Starting to process dataset...")
+            os.makedirs(
+                f"{self.config.dataset_config.output_dir}_{self.config.dataset_config.dataset_name}_{self.config.dataset_config.uuid}",
+                exist_ok=True
+            )
+
+            if isinstance(dataset, IterableDataset):
+                return self.process_streaming_dataset(dataset)
+            else:
+                return self.process_regular_dataset(dataset)
+        except Exception as e:
```

Comment:
in future, it would be great to cache or dump the failing examples for future debugging at that point.